Some `FormulaEngine` tests are flaky

### What happened?

I got at least this test failing: `TestFormulaComposition.test_formula_composition`.

### What did you expect instead?

Tests pass.

### Affected version(s)

_No response_

### Affected part(s)

Data pipeline (part:data-pipeline)

### Extra information

Log in: https://github.com/frequenz-floss/frequenz-sdk-python/actions/runs/4293183301/jobs/7480543225

```
=================================== FAILURES ===================================
_______________ TestFormulaComposition.test_formula_composition ________________

self = <tests.timeseries._formula_engine.test_formula_composition.TestFormulaComposition object at 0x7f43af5e2830>
mocker = <pytest_mock.plugin.MockerFixture object at 0x7f43aeb86f50>

    async def test_formula_composition(  # pylint: disable=too-many-locals
        self,
        mocker: MockerFixture,
    ) -> None:
        """Test the composition of formulas."""
        mockgrid = MockMicrogrid(grid_side_meter=False, sample_rate_s=0.05)
        mockgrid.add_batteries(3)
        mockgrid.add_solar_inverters(2)
        request_chan, channel_registry = await mockgrid.start(mocker)
        logical_meter = LogicalMeter(
            channel_registry,
            request_chan.new_sender(),
            microgrid.get().component_graph,
        )
    
        main_meter_recv = await get_resampled_stream(
            logical_meter,
            channel_registry,
            request_chan.new_sender(),
            4,
            ComponentMetricId.ACTIVE_POWER,
        )
        grid_power_recv = await logical_meter.grid_power()
        battery_power_recv = await logical_meter.battery_power()
        pv_power_recv = await logical_meter.pv_power()
    
        engine = (pv_power_recv.clone() + battery_power_recv.clone()).build("inv_power")
        inv_calc_recv = engine.new_receiver()
    
        count = 0
        for _ in range(10):
            grid_pow = await grid_power_recv.receive()
            pv_pow = await pv_power_recv.receive()
            bat_pow = await battery_power_recv.receive()
            main_pow = await main_meter_recv.receive()
            inv_calc_pow = await inv_calc_recv.receive()
    
>           assert grid_pow is not None and grid_pow.value is not None
E           assert (Sample(timestamp=datetime.datetime(2023, 2, 28, 13, 2, 47, 135347, tzinfo=datetime.timezone.utc), value=None) is not None and None is not None)
E            +  where None = Sample(timestamp=datetime.datetime(2023, 2, 28, 13, 2, 47, 135347, tzinfo=datetime.timezone.utc), value=None).value

tests/timeseries/_formula_engine/test_formula_composition.py:60: AssertionError
```

Another instance where it happened:

```
=================================== FAILURES ===================================
_______________ TestFormulaComposition.test_formula_composition ________________

self = <tests.timeseries._formula_engine.test_formula_composition.TestFormulaComposition object at 0x7f229e2c5130>
mocker = <pytest_mock.plugin.MockerFixture object at 0x7f229db98910>

    async def test_formula_composition(  # pylint: disable=too-many-locals
        self,
        mocker: MockerFixture,
    ) -> None:
        """Test the composition of formulas."""
        mockgrid = MockMicrogrid(grid_side_meter=False, sample_rate_s=0.05)
        mockgrid.add_batteries(3)
        mockgrid.add_solar_inverters(2)
        request_chan, channel_registry = await mockgrid.start(mocker)
        logical_meter = LogicalMeter(
            channel_registry,
            request_chan.new_sender(),
            microgrid.get().component_graph,
        )
    
        main_meter_recv = await get_resampled_stream(
            logical_meter,
            channel_registry,
            request_chan.new_sender(),
            4,
            ComponentMetricId.ACTIVE_POWER,
        )
        grid_power_recv = await logical_meter.grid_power()
        battery_power_recv = await logical_meter.battery_power()
        pv_power_recv = await logical_meter.pv_power()
    
        engine = (pv_power_recv.clone() + battery_power_recv.clone()).build("inv_power")
        inv_calc_recv = engine.new_receiver()
    
        count = 0
        for _ in range(10):
            grid_pow = await grid_power_recv.receive()
            pv_pow = await pv_power_recv.receive()
            bat_pow = await battery_power_recv.receive()
            main_pow = await main_meter_recv.receive()
            inv_calc_pow = await inv_calc_recv.receive()
    
>           assert grid_pow is not None and grid_pow.value is not None
E           assert (Sample(timestamp=datetime.datetime(2023, 2, 28, 18, 10, 24, 253714, tzinfo=datetime.timezone.utc), value=None) is not None and None is not None)
E            +  where None = Sample(timestamp=datetime.datetime(2023, 2, 28, 18, 10, 24, 253714, tzinfo=datetime.timezone.utc), value=None).value

tests/timeseries/_formula_engine/test_formula_composition.py:60: AssertionError
```
Link: https://github.com/frequenz-floss/frequenz-sdk-python/actions/runs/4295981776/jobs/7487129493

Another instance, this time with a different assert failure:

```
=================================== FAILURES ===================================
__________________________ TestLogicalMeter.test_soc ___________________________

self = <tests.timeseries.test_logical_meter.TestLogicalMeter object at 0x7f64f654f5d0>
mocker = <pytest_mock.plugin.MockerFixture object at 0x7f64efc7d550>

    async def test_soc(self, mocker: MockerFixture) -> None:
        """Test the soc calculation."""
        mockgrid = MockMicrogrid(grid_side_meter=False, sample_rate_s=0.02)
        mockgrid.add_solar_inverters(2)
        mockgrid._id_increment = 8  # pylint: disable=protected-access
        mockgrid.add_batteries(3)
        request_chan, channel_registry = await mockgrid.start(mocker)
        logical_meter = LogicalMeter(
            channel_registry,
            request_chan.new_sender(),
            microgrid.get().component_graph,
        )
    
        soc_recv = await logical_meter._soc()  # pylint: disable=protected-access
    
        bat_receivers = [
            await get_resampled_stream(
                logical_meter,
                channel_registry,
                request_chan.new_sender(),
                bat_id,
                ComponentMetricId.SOC,
            )
            for bat_id in mockgrid.battery_ids
        ]
    
        await synchronize_receivers([soc_recv, *bat_receivers])
    
        for ctr in range(10):
            bat_vals = []
            for recv in bat_receivers:
                val = await recv.receive()
                assert val is not None and val.value is not None
                bat_vals.append(val.value)
    
            assert len(bat_vals) == 3
            # After 5 values, the inverter with component_id > 100 stops sending
            # data. And the values from the last battery goes out of the calculation.
            # So we drop it from out control value as well.
            if ctr >= 5:
                bat_vals = bat_vals[:2]
    
            soc_sample = await soc_recv.receive()
            assert soc_sample is not None and soc_sample.value is not None
    
>           assert isclose(soc_sample.value, sum(bat_vals) / len(bat_vals))
E           assert False
E            +  where False = isclose(15.0, (46.5 / 3))
E            +    where 15.0 = Sample(timestamp=datetime.datetime(2023, 3, 1, 10, 36, 4, 580789, tzinfo=datetime.timezone.utc), value=15.0).value
E            +    and   46.5 = sum([14.5, 15.5, 16.5])
E            +    and   3 = len([14.5, 15.5, 16.5])

tests/timeseries/test_logical_meter.py:235: AssertionError
```
Link: https://github.com/frequenz-floss/frequenz-sdk-python/actions/runs/4302570564/jobs/7501231010

<details>
<summary>Another instance</summary>

```
=================================== FAILURES ===================================
__________________________ TestLogicalMeter.test_soc ___________________________

self = <tests.timeseries.test_logical_meter.TestLogicalMeter object at 0x7f17e1716a10>
mocker = <pytest_mock.plugin.MockerFixture object at 0x7f17daf40d50>

    async def test_soc(self, mocker: MockerFixture) -> None:
        """Test the soc calculation."""
        mockgrid = MockMicrogrid(grid_side_meter=False, sample_rate_s=0.02)
        mockgrid.add_solar_inverters(2)
        mockgrid._id_increment = 8  # pylint: disable=protected-access
        mockgrid.add_batteries(3)
        request_chan, channel_registry = await mockgrid.start(mocker)
        logical_meter = LogicalMeter(
            channel_registry,
            request_chan.new_sender(),
            microgrid.get().component_graph,
        )
    
        soc_recv = await logical_meter._soc()  # pylint: disable=protected-access
    
        bat_receivers = [
            await get_resampled_stream(
                logical_meter,
                channel_registry,
                request_chan.new_sender(),
                bat_id,
                ComponentMetricId.SOC,
            )
            for bat_id in mockgrid.battery_ids
        ]
    
        await synchronize_receivers([soc_recv, *bat_receivers])
    
        for ctr in range(10):
            bat_vals = []
            for recv in bat_receivers:
                val = await recv.receive()
                assert val is not None and val.value is not None
                bat_vals.append(val.value)
    
            assert len(bat_vals) == 3
            # After 5 values, the inverter with component_id > 100 stops sending
            # data. And the values from the last battery goes out of the calculation.
            # So we drop it from out control value as well.
            if ctr >= 5:
                bat_vals = bat_vals[:2]
    
            soc_sample = await soc_recv.receive()
            assert soc_sample is not None and soc_sample.value is not None
    
>           assert isclose(soc_sample.value, sum(bat_vals) / len(bat_vals))
E           assert False
E            +  where False = isclose(15.5, (47.5 / 3))
E            +    where 15.5 = Sample(timestamp=datetime.datetime(2023, 3, 2, 8, 19, 27, 286960, tzinfo=datetime.timezone.utc), value=15.5).value
E            +    and   47.5 = sum([15.0, 16.0, 16.5])
E            +    and   3 = len([15.0, 16.0, 16.5])

tests/timeseries/test_logical_meter.py:235: AssertionError
```

</details>

https://github.com/leandro-lucarella-frequenz/frequenz-sdk-python/actions/runs/4311908197/jobs/7521820394#step:6:632

Python 3.11:
https://github.com/frequenz-floss/frequenz-sdk-python/actions/runs/4312050797/jobs/7522133295
```
FAILED tests/timeseries/test_logical_meter.py::TestLogicalMeter::test_soc - assert False
 +  where False = isclose(15.0, (46.5 / 3))
 +    where 15.0 = Sample(timestamp=datetime.datetime(2023, 3, 2, 8, 38, 33, 509482, tzinfo=datetime.timezone.utc), value=15.0).value
 +    and   46.5 = sum([14.5, 15.5, 16.5])
 +    and   3 = len([14.5, 15.5, 16.5])
```

--- a/tests/timeseries/mock_microgrid.py
+++ b/tests/timeseries/mock_microgrid.py
@@ -75,6 +75,7 @@ class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
         self.battery_ids: list[int] = []
         self.evc_ids: list[int] = []
         self.meter_ids: list[int] = [4]
+        self.bat_inv_map: dict[int, int] = {}
 
         self.evc_component_states: dict[int, EVChargerComponentState] = {}
         self.evc_cable_states: dict[int, EVChargerCableState] = {}
@@ -179,6 +180,7 @@ class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
             self.meter_ids.append(meter_id)
             self.battery_inverter_ids.append(inv_id)
             self.battery_ids.append(bat_id)
+            self.bat_inv_map[bat_id] = inv_id
 
             self._components.add(
                 Component(

--- a/tests/timeseries/test_logical_meter.py
+++ b/tests/timeseries/test_logical_meter.py
@@ -20,6 +20,8 @@ from ._formula_engine.utils import (
 )
 from .mock_microgrid import MockMicrogrid
 
+# pylint: disable=too-many-locals
+
 
 class TestLogicalMeter:
     """Tests for the logical meter."""
@@ -202,8 +204,8 @@ class TestLogicalMeter:
 
         soc_recv = await logical_meter._soc()  # pylint: disable=protected-access
 
-        bat_receivers = [
-            await get_resampled_stream(
+        bat_receivers = {
+            bat_id: await get_resampled_stream(
                 logical_meter,
                 channel_registry,
                 request_chan.new_sender(),
@@ -211,23 +213,38 @@ class TestLogicalMeter:
                 ComponentMetricId.SOC,
             )
             for bat_id in mockgrid.battery_ids
-        ]
+        }
 
-        await synchronize_receivers([soc_recv, *bat_receivers])
+        bat_inv_map = mockgrid.bat_inv_map
+        inv_receivers = {
+            inverter_id: await get_resampled_stream(
+                logical_meter,
+                channel_registry,
+                request_chan.new_sender(),
+                inverter_id,
+                ComponentMetricId.ACTIVE_POWER,
+            )
+            for inverter_id in mockgrid.bat_inv_map.values()
+        }
 
-        for ctr in range(10):
+        await synchronize_receivers(
+            [soc_recv, *bat_receivers.values(), *inv_receivers.values()]
+        )
+
+        for _ in range(10):
             bat_vals = []
-            for recv in bat_receivers:
-                val = await recv.receive()
+            for bat_id, bat_recv in bat_receivers.items():
+                inv_id = bat_inv_map[bat_id]
+                inv_recv = inv_receivers[inv_id]
+                inv_msg = await inv_recv.receive()
+                # We won't use batteries where adjacent inverter is not sending active
+                # power.
+                if inv_msg.value is None:
+                    continue
+
+                val = await bat_recv.receive()
                 assert val is not None and val.value is not None
                 bat_vals.append(val.value)
-
-            assert len(bat_vals) == 3
-            # After 5 values, the inverter with component_id > 100 stops sending
-            # data. And the values from the last battery goes out of the calculation.
-            # So we drop it from out control value as well.
-            if ctr >= 5:
-                bat_vals = bat_vals[:2]
 
             soc_sample = await soc_recv.receive()
             assert soc_sample is not None and soc_sample.value is not None

